### ISSUE TITLE
sql: Initial support for PostgREST

### ISSUE THREAD

**Issue by awoods187** [cockroachdb/cockroach#69010]
Add support for [PostgREST](https://postgrest.org/en/v8.0/). PostgREST is a standalone web server that turns your PostgreSQL database directly into a RESTful API. The structural constraints and permissions in the database determine the API endpoints and operations.

@nvanbenschoten did an early investigation and identified the following items needed for support:

Done:
 
- [x] regrole: https://github.com/cockroachdb/cockroach/pull/68877
- [x] SET LOCAL: https://github.com/cockroachdb/cockroach/issues/32562
- [x] SET ROLE: https://github.com/cockroachdb/cockroach/issues/15005
- [x] builtin: pg_is_other_temp_schema: https://github.com/cockroachdb/cockroach/pull/69909
- [x] builtin: pg_has_role: https://github.com/cockroachdb/cockroach/pull/69939
- [x] builtin: information_schema._pg_char_max_length: https://github.com/cockroachdb/cockroach/pull/69913
- [x] builtin: information_schema._pg_index_position: https://github.com/cockroachdb/cockroach/pull/69911
- [x] builtin: information_schema._pg_truetypid: https://github.com/cockroachdb/cockroach/pull/69913
- [x] builtin: information_schema._pg_truetypmod: https://github.com/cockroachdb/cockroach/pull/69913
- [x] support preferred binary operators: https://github.com/cockroachdb/cockroach/pull/70066
- [x] builtin: array_agg(jsonb): https://github.com/cockroachdb/cockroach/pull/70041
- [x] DistSQL loses named tuple field names: https://github.com/cockroachdb/cockroach/issues/70360
- [x] builtin: array_agg(record): https://github.com/cockroachdb/cockroach/issues/70029
- [x] create named composite type per table: https://github.com/cockroachdb/cockroach/issues/70036
- [x] CTE source not available as function argument in JOIN target: https://github.com/cockroachdb/cockroach/issues/70342
- [x] RECURSE UNION: https://github.com/cockroachdb/cockroach/issues/46642
- [x] builtin: json_populate_record, json_populate_recordset: https://github.com/cockroachdb/cockroach/pull/70115
- [x] `T_unknown` wire decoding: https://github.com/cockroachdb/cockroach/issues/70027
- [x] support string `literal -> tuple[]` parsing: https://github.com/cockroachdb/cockroach/issues/70030
- [x] support customized session options: https://github.com/cockroachdb/cockroach/issues/70033
- [x] grant CONNECT privilege to PUBLIC role on new databases: https://github.com/cockroachdb/cockroach/issues/70098. Blocking OpenAPI: https://github.com/cockroachdb/cockroach/issues/69010#issuecomment-917838859. Can work around with manual `GRANT CONNECT ON DATABASE <db> TO PUBLIC`.

Hard blockers:

None

Soft blockers (workarounds available or only blocking limited functionality):
- [x] builtin: json_to_record, json_to_recordset: https://github.com/cockroachdb/cockroach/issues/70037. Needed to call stored procedures. But CRDB does not support stored procedures. So not immediately blocking anything.

gz#9915

Jira issue: CRDB-9405

**Comment by otan** [comment]
do you know why postgrest needs SET ROLE?

**Comment by nvanbenschoten** [comment]
https://postgrest.org/en/stable/auth.html does a better job explaining this than I will be able to. There is overlap here with `ROW LEVEL SECURITY`, but I would not consider that to be a blocker for PostgREST support.

**Comment by nvanbenschoten** [comment]
I've knocked out each of the missing introspection builtins. They were all reasonably straightforward, in part due to recent investments in role-based access control, and will land in v22.1. With that added PG compatibility, we're now able to run startup queries like the following from in PostgREST:

- https://github.com/PostgREST/postgrest/blob/b05898d17f8e33c8c82fc1d05a30eb3044999668/src/PostgREST/DbStructure.hs#L538
- https://github.com/PostgREST/postgrest/blob/b05898d17f8e33c8c82fc1d05a30eb3044999668/src/PostgREST/DbStructure.hs#L709

From what I can tell, that leaves [WITH RECURSIVE UNION](https://github.com/cockroachdb/cockroach/issues/46642) as the remaining startup blocker. @RaduBerinde has estimated that will take 2-4 weeks of work to land.

Things aren't as far along with the remaining runtime blockers. I've identified 4 cases where our JSON support is insufficient to run queries through PostgREST. The concerning part about these is that they aren't surface-level deficiencies. From what I can tell, our SQL type-system is not currently powerful enough to implement builtins like `json_populate_recordset` and `json_to_recordset`. This is related to limitations in our handling of `record` types. Unlike Postgres, we don't automatically define a named record type associated with the structure of each table. This is necessary because uses of `json_populate_recordset` look like `json_populate_recordset(null::mytable, '[{"a":1,"b":2}, {"a":3,"b":4}]')`. In fact, there appear to be gaps around our handling of named record types in general. We also lack the ability to define the structure of a record with an AS clause, which is necessary for `json_to_recordset`, which is used like `json_to_recordset('[{"a":1,"b":"foo"}, {"a":"2","c":"bar"}]') as x(a int, b text)`. These are important because this mapping from JSONB -> record is how PostgREST and other layers like it (at least [Hasura](https://github.com/hasura/graphql-engine/blob/7ca48decfb3718301a2fcdf4515db70fc6f0d05d/server/src-lib/Hasura/GraphQL/Execute/Action.hs#L256), probably others) perform type-safe mutations into structured relatioanl tables.

I'd like to get @otan's eyes on this at some point because he'll be able to classify these limitations and map them to existing initiatives.

Until then, my takeaway is that we need to keep investing in our type system and JSON support. More and more tools are interested in tight JSON integration, and we have some real gaps in this area. These gaps will be considered hard blockers for these tools, and they're not the kinds of blockers that can be worked around in the tools. They need to be solved in CockroachDB itself.

**Comment by rafiss** [comment]
There have been a few other issues that have come up around our divergence in how we handle `record` types. A lot of it comes down to needing to more closely implement this: https://www.postgresql.org/docs/12/rowtypes.html#ROWTYPES-USAGE

- resolving table names and tuples (https://github.com/cockroachdb/cockroach/issues/26719 and https://github.com/cockroachdb/cockroach/issues/45854) - fixed in 21.2
- equivalence of `composite_value` and `composite_value.*` (https://github.com/cockroachdb/cockroach/issues/60549) - fixed in 21.2
- functional notation for composite values (https://github.com/cockroachdb/cockroach/issues/47477) - unresolved

just providing these for context - perhaps the issues that you discuss above should be included as part of this theme.

**Comment by nvanbenschoten** [comment]
By the way, [Supabase sits on top of PostgREST](https://supabase.io/docs/architecture), so for all intents and purposes, these are blockers for Supabase support as well.

**Comment by nvanbenschoten** [comment]
With the new PG compatibility added this week, I took another pass at running PostgREST on CockroachDB. This revealed a few blockers I had missed before, which I've filed issues about and added to the list above. I then did my best to aggressively hack around them in a prototype branch. I was able hack around all but the last two blockers (#70036 and #70037) in this branch: https://github.com/nvanbenschoten/cockroach/commits/nvanbenschoten/postgrest.

As it turns out, this was enough to get PostgREST up, running, and partially functional! With these changes, read-only queries (GET requests) are supported, as are delete statements (DELETE requests). However, INSERT/UPDATE/UPSERT statements (POST and PUT requests) are still blocked on those two remaining items.

This is enough to give a taste of PostgREST. What follows was adapted from [the PostgREST tutorial](https://postgrest.org/en/v8.0/tutorials/tut0.html), adapted slightly for the `movr` schema.

### run cockroachdb
```
➜ ./cockroach demo --insecure
```

### setup roles
```sql
create role web_anon nologin;
grant select on all tables in schema public to web_anon;

create role authenticator login;
grant web_anon to authenticator;
```

### configure postgrest
```
echo 'db-uri = "postgres://authenticator@localhost:26257/movr"' >> tutorial.conf
echo 'db-schema = "public"'                                     >> tutorial.conf
echo 'db-anon-role = "web_anon"'                                >> tutorial.conf
```

### run postgrest
```
➜ ./postgrest tutorial.conf
```

### issue GET requests
```
➜ curl -s 'http://localhost:3000/users?limit=3' | jq
[
  {
    "address": "88194 Angela Gardens Suite 94",
    "city": "amsterdam",
    "credit_card": "4443538758",
    "id": "ae147ae1-47ae-4800-8000-000000000022",
    "name": "Tyler Dalton"
  },
  {
    "address": "29590 Butler Plain Apt. 25",
    "city": "amsterdam",
    "credit_card": "3750897994",
    "id": "b3333333-3333-4000-8000-000000000023",
    "name": "Dillon Martin"
  },
  {
    "address": "32768 Eric Divide Suite 88",
    "city": "amsterdam",
    "credit_card": "8107478823",
    "id": "b851eb85-1eb8-4000-8000-000000000024",
    "name": "Deborah Carson"
  }
]

➜ curl -s 'http://localhost:3000/vehicles?status=eq.in_use&select=city,ext&limit=2' | jq
[
  {
    "city": "amsterdam",
    "ext": {
      "color": "red"
    }
  },
  {
    "city": "boston",
    "ext": {
      "color": "blue"
    }
  }
]
```

### issue DELETE requests
```
## count rows before
➜ curl -s -I 'http://localhost:3000/vehicle_location_histories' -H 'Prefer: count=exact'
HTTP/1.1 200 OK
Date: Fri, 10 Sep 2021 16:57:21 GMT
Server: postgrest/20210706 (0961a58)
Content-Range: 0-999/1000
Content-Location: /vehicle_location_histories
Content-Type: application/json; charset=utf-8

## try to delete rows in boston
➜ curl -s -X DELETE 'http://localhost:3000/vehicle_location_histories?city=eq.boston' | jq
{
  "hint": null,
  "details": null,
  "code": "42501",
  "message": "user web_anon does not have DELETE privilege on relation vehicle_location_histories"
}

## grant permissions
grant delete on all tables in schema public to web_anon;

## try to delete 5 rows again
➜ curl -s -X DELETE 'http://localhost:3000/vehicle_location_histories?city=eq.boston'

## count rows after
➜ curl -s -I 'http://localhost:3000/vehicle_location_histories' -H 'Prefer: count=exact'
HTTP/1.1 200 OK
Date: Fri, 10 Sep 2021 17:04:23 GMT
Server: postgrest/20210706 (0961a58)
Content-Range: 0-888/889
Content-Location: /vehicle_location_histories
Content-Type: application/json; charset=utf-8
```

### issue POST request
```
## grant permissions
grant insert on all tables in schema public to web_anon;

➜ curl -s 'http://localhost:3000/promo_codes' -X POST \
     -H "Content-Type: application/json" \
     -d '{"code": "time to REST"}' | jq
{
  "hint": "You have attempted to use a feature that is not yet implemented.\nSee: https://go.crdb.dev/issue-v/33285/v21.2",
  "details": null,
  "code": "0A000",
  "message": "json_populate_recordset(): unimplemented: this function is not yet supported"
}
```

**Comment by jordanlewis** [comment]
Nice going!

**Comment by rafiss** [comment]
This is awesome!

I wonder if we could start off a conversation about adding "official" CockroachDB support to PostgREST. That way the functionality would not be blocked on resolving some of these deep and thorny compat issues in CRDB. I do understand from the summary above that some of the issues are hard blockers that can't be worked around. But it couldn't hurt to have more possible approaches to resolving the other issues.

**Comment by vy-ton** [comment]
> I wonder if we could start off a conversation about adding "official" CockroachDB support to PostgREST.

I can follow up here

**Comment by nvanbenschoten** [comment]
With the bug fix to our access privilege inquiry functions (i.e. the `has_foo_privilege` builtins) in https://github.com/cockroachdb/cockroach/pull/69939#issuecomment-917785405 and after running `grant connect on database movr to public` to work around https://github.com/cockroachdb/cockroach/issues/70098, PostgREST's [OpenAPI](https://www.openapis.org/) implementation is now supported as well. This makes it possible to introspect the system and query the list of all endpoints and parameters, similar to the "self-documenting" behavior of GraphQL.

It also means that we can put a Swagger UI in front and interact with the API. It's neat!

![localhost_3200_](https://user-images.githubusercontent.com/5438456/133025269-fe95997d-06ec-48d4-b18a-4bd8b7a8781e.png)

Also, we can use one of the [PostgREST client libraries](https://postgrest.org/en/v8.0/ecosystem.html#client-side-libraries) to trivially query the database. No connection pools. No state. I'm beginning to remember why people like HTTP.

```js
import { PostgrestClient } from '@supabase/postgrest-js'

const REST_URL = 'http://localhost:3000'
const postgrest = new PostgrestClient(REST_URL)

const { data: rides, error } = await postgrest
	.from('rides')
	.select()
	.eq('id', '54fdf3b6-45a1-4c00-8000-0000000000a6');

console.log(rides);
```

**Comment by nvanbenschoten** [comment]
We've landed a few more changes on master since the last progress update here. @jordanlewis has also made some very nice progress on the two issues that were blocking INSERTs/UPDATEs with https://github.com/cockroachdb/cockroach/pull/70115 and https://github.com/cockroachdb/cockroach/pull/70100. I've rebased the [prototype branch](https://github.com/nvanbenschoten/cockroach/commits/nvanbenschoten/postgrest) from earlier onto master and cherry-picked those two PRs on top of it. It looks like that was enough to unblock [INSERTs and UPDATEs](https://postgrest.org/en/v8.0/api.html#insertions-updates)!

```sh
# INSERT
➜ curl -s 'http://localhost:3000/promo_codes?code=like.%torest' | jq
[]

➜ curl -s -X POST 'http://localhost:3000/promo_codes' \
     -H "Content-Type: application/json" \
     -d '{"code": "timetorest", "description": "newly created", "rules": {"details": "lol, this is json"}}' | jq

➜ curl -s 'http://localhost:3000/promo_codes?code=like.%torest' | jq
[
  {
    "code": "timetorest",
    "creation_time": null,
    "description": "newly created",
    "expiration_time": null,
    "rules": {
      "details": "lol, this is json"
    }
  }
]

# UPDATE
➜ curl -s -X PATCH 'http://localhost:3000/promo_codes?code=like.%torest' \
     -H "Content-Type: application/json" \
     -d '{"description": "created and updated"}' | jq

➜ curl -s 'http://localhost:3000/promo_codes?code=like.%torest' | jq
[
  {
    "code": "timetorest",
    "creation_time": null,
    "description": "created and updated",
    "expiration_time": null,
    "rules": {
      "details": "lol, this is json"
    }
  }
]
```

I did find [one new issue](https://github.com/cockroachdb/cockroach/issues/70342) that blocks `Prefer: return=headers-only` and `Prefer: return=representation` HTTP header support for POST and PATCH requests (similar to `RETURNING id` and `RETURNING *`, respectively). But this isn't a hard blocker.

[Bulk inserts](https://postgrest.org/en/v8.0/api.html#bulk-insert) also work.

As do [both forms of UPSERTs](https://postgrest.org/en/v8.0/api.html#upsert):
```sh
➜ curl -s -X POST 'http://localhost:3000/promo_codes' \
     -H "Content-Type: application/json" \
     -H "Prefer: resolution=merge-duplicates" \
     -d '{"code": "timetorest", "description": "created, updated, upserted"}' | jq

➜ curl -s 'http://localhost:3000/promo_codes?code=like.%torest' | jq
[
  {
    "code": "timetorest",
    "creation_time": null,
    "description": "created, updated, upserted",
    "expiration_time": null,
    "rules": {
      "details": "lol, this is json"
    }
  }
]

➜ curl -s -X PUT 'http://localhost:3000/promo_codes?code=eq.timetorest' \
     -H "Content-Type: application/json" \
     -d '{"code": "timetorest", "description": "created, updated, upserted 2x"}' | jq

➜ curl -s 'http://localhost:3000/promo_codes?code=like.%torest' | jq
[
  {
    "code": "timetorest",
    "creation_time": null,
    "description": "created, updated, upserted 2x",
    "expiration_time": null,
    "rules": {
      "details": "lol, this is json"
    }
  }
]
```

Also, one thing I meant to show before is that JOINs work and they're pretty convenient to use:
```sh
➜ curl -s 'http://localhost:3000/rides?limit=1&select=*' | jq
[
  {
    "city": "amsterdam",
    "end_address": "66037 Belinda Plaza Apt. 93",
    "end_time": "2018-12-14T08:04:05",
    "id": "ab020c49-ba5e-4800-8000-00000000014e",
    "revenue": 77,
    "rider_id": "c28f5c28-f5c2-4000-8000-000000000026",
    "start_address": "1905 Christopher Locks Apt. 77",
    "start_time": "2018-12-13T03:04:05",
    "vehicle_city": "amsterdam",
    "vehicle_id": "aaaaaaaa-aaaa-4800-8000-00000000000a"
  }
]

➜ curl -s 'http://localhost:3000/rides?limit=1&select=*,vehicles(*),users(*)' | jq
[
  {
    "city": "amsterdam",
    "end_address": "66037 Belinda Plaza Apt. 93",
    "end_time": "2018-12-14T08:04:05",
    "id": "ab020c49-ba5e-4800-8000-00000000014e",
    "revenue": 77,
    "rider_id": "c28f5c28-f5c2-4000-8000-000000000026",
    "start_address": "1905 Christopher Locks Apt. 77",
    "start_time": "2018-12-13T03:04:05",
    "users": {
      "address": "14729 Karen Radial",
      "city": "amsterdam",
      "credit_card": "5844236997",
      "id": "c28f5c28-f5c2-4000-8000-000000000026",
      "name": "Maria Weber"
    },
    "vehicle_city": "amsterdam",
    "vehicle_id": "aaaaaaaa-aaaa-4800-8000-00000000000a",
    "vehicles": {
      "city": "amsterdam",
      "creation_time": "2019-01-02T03:04:05",
      "current_location": "62609 Stephanie Route",
      "ext": {
        "color": "red"
      },
      "id": "aaaaaaaa-aaaa-4800-8000-00000000000a",
      "owner_id": "c28f5c28-f5c2-4000-8000-000000000026",
      "status": "in_use",
      "type": "scooter"
    }
  }
]
```

So we're actually pretty close here. I think the key is that PostgREST isn't trying to perform schema changes or make any system catalog modifications itself. It's actually entirely read-only outside of user-performed mutations, which is a pretty respectable stance to take. I wish more tools embraced that idea. What this means for Cockroach is that it's not pushing quite as hard on advanced features like user-defined types or user-defined functions as other similar tools do (although you can certainly [use your own](https://postgrest.org/en/v8.0/api.html#stored-procedures) if you want).

What PostgREST does push pretty hard on is roles and permissions, because it delegates all [authorization to the database](https://postgrest.org/en/v8.0/auth.html). This is pretty critical if you intend to actually expose a service like this on the open web because it means that your database's access control policy is the only thing keeping your data safe. I say this mostly to point out that the recent work by @RichardJCai and @rafiss in this area has made a big impact for PostgREST support, even if we didn't realize it beforehand. I imagine the blocker list would have been 50% longer if we did this investigation 6 months ago.

**Comment by jordanlewis** [comment]
So cool 😍

**Comment by jordanlewis** [comment]
`json_to_record` support is coming here: #82435. @nvanbenschoten, do you remember what was still blocked behind that one?

**Comment by nvanbenschoten** [comment]
I think this was needed to call stored procedures. See [`requestToCallProcQuery`](https://cs.github.com/PostgREST/postgrest/blob/fe8b32d0775dd7e4a1b0e26dee713fd3e17c7f1e/src/PostgREST/Query/QueryBuilder.hs?q=json_to_record#L189). There are no other uses of `json_to_record` or `jsonb_to_record` in postgrest.

**Comment by dikshant** [comment]
With 22.2 we will have initial support for PostgREST which enables basic CRUD operations on REST APIs. 
The remaining work this will be tracked here:
https://github.com/cockroachdb/cockroach/issues/91331
This issue will hence be closed.

**Comment by alexsala** [comment]
Hello! 
Following a slack conversation with @alyshanjahani-crl, I'm appending to this issue as CRUD seems to fail on the following example derived from @nvanbenschoten 's work


#### Context 
Mac M1 Pro - MacOS 14.2.1

postgREST version 12.0.2

cockroach version: 
- Build Tag:        v23.1.14
- Build Time:       2024/01/16 13:07:29
- Distribution:     CCL
- Platform:         darwin arm64 (aarch64-apple-darwin21.2)

#### install & run cockroach
```
brew install cockroachdb/tap/cockroach
cockroach demo --insecure
```

#### Setup roles
```sql
create role web_anon nologin;
grant select on all tables in schema public to web_anon;
create role authenticator login;
grant web_anon to authenticator;
```

#### Install, configure & run postgREST
```bash
brew install postgrest
echo 'db-uri = "postgres://authenticator@localhost:26257/movr"' >> tutorial.conf
echo 'db-schema = "public"'                                     >> tutorial.conf
echo 'db-anon-role = "web_anon"'                                >> tutorial.conf
postgrest tutorial.conf
```

#### GET query
```bash
curl -s 'http://localhost:3000/vehicles?status=eq.in_use&select=city,ext&limit=2'
```

This returns the right result, but add an extra `^A` character at the start.

#### POST query
```bash
curl -s -X POST 'http://localhost:3000/promo_codes' \
     -H "Content-Type: application/json" \
     -d '{"code": "timetorest", "description": "newly created", "rules": {"details": "lol, this is json"}}' 
```

This returns `{}` and no insert was done.

**Comment by dikshant** [comment]
@alexsala You will want to try Postgres 10.0. Newer versions do not work.

**Comment by alexsala** [comment]
Hi @dikshant, thanks for letting me know. Will rollback to v10 and see if it all works.
